rust/src/shards/gui/widgets/image_util.rs

```diff
@@ -0,0 +1,104 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use crate::fourCharacterCode;
+use crate::shardsc::gfx_TexturePtr;
+use crate::shardsc::gfx_TexturePtr_getResolution_ext;
+use crate::shardsc::SHImage;
+use crate::shardsc::SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA;
+use crate::types::common_type;
+use crate::types::ParamVar;
+use crate::types::Type;
+use crate::types::Var;
+use crate::types::FRAG_CC;
+use std::ptr::null_mut;
+
+pub const TEXTURE_CC: i32 = fourCharacterCode(*b"tex_");
+
+lazy_static! {
+  pub static ref TEXTURE_TYPE: Type = Type::object(FRAG_CC, TEXTURE_CC);
+  pub static ref TEXTURE_OR_IMAGE_TYPES: Vec<Type> = vec![common_type::image, *TEXTURE_TYPE];
+}
+
+pub fn get_scale(scale_var: &ParamVar) -> Result<egui::Vec2, &'static str> {
+  let scale: (f32, f32) = scale_var.get().try_into()?;
+  Ok(egui::vec2(scale.0, scale.1))
+}
+
+pub struct CachedUIImage {
+  texture_handle: Option<egui::TextureHandle>,
+  prev_ptr: *mut u8,
+}
+
+impl Default for CachedUIImage {
+  fn default() -> Self {
+    Self {
+      texture_handle: None,
+      prev_ptr: null_mut(),
+    }
+  }
+}
+
+impl CachedUIImage {
+  pub fn get_egui_texture_from_image<'a>(
+    &'a mut self,
+    input: &Var,
+    ui: &egui::Ui,
+  ) -> Result<&'a egui::TextureHandle, &'static str> {
+    let shimage: &SHImage = input.try_into()?;
+    let ptr = shimage.data;
+    Ok(if ptr != self.prev_ptr {
+      let image: egui::ColorImage = shimage.into();
+      self.prev_ptr = ptr;
+      self.texture_handle.insert(ui.ctx().load_texture(
+        format!("UI.Image: {:p}", shimage.data),
+        image,
+        Default::default(),
+      ))
+    } else {
+      self.texture_handle.as_ref().unwrap()
+    })
+  }
+}
+
+pub fn get_egui_texture_from_gfx(
+  input: &Var,
+) -> Result<(egui::TextureId, egui::Vec2), &'static str> {
+  let texture_ptr: *mut gfx_TexturePtr =
+    Var::from_object_ptr_mut_ref::<gfx_TexturePtr>(*input, &TEXTURE_TYPE)?;
+  let texture_size = {
+    let texture_res = unsafe { gfx_TexturePtr_getResolution_ext(texture_ptr) };
+    egui::vec2(texture_res.x as f32, texture_res.y as f32)
+  };
+
+  Ok((
+    egui::epaint::TextureId::User(texture_ptr as u64),
+    texture_size,
+  ))
+}
+
+impl From<&SHImage> for egui::ColorImage {
+  fn from(image: &SHImage) -> egui::ColorImage {
+    assert_eq!(image.channels, 4);
+
+    let size = [image.width as _, image.height as _];
+    let rgba = unsafe {
+      core::slice::from_raw_parts(
+        image.data,
+        image.width as usize * image.channels as usize * image.height as usize,
+      )
+    };
+
+    if image.flags & SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA as u8
+      == SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA as u8
+    {
+      let pixels = rgba
+        .chunks_exact(4)
+        .map(|p| egui::Color32::from_rgba_premultiplied(p[0], p[1], p[2], p[3]))
+        .collect();
```

Comment:
Allocation, write it simple.
Those things are like `linq` sadly...

Reply:
I thought rust compiler was optimizing them away. I saw a few posts on reddit where it showed that the resulting assembly in release was optimal in most cases.

Reply:
Not too much point in rewriting it, I think we can merge it later with what's being done by GFX.Texture (less image conversions/copies)

Reply:
I think it's good preventive optimisation here. We should have some hard rules, e.g. no `collect`.
Btw not sure how compiler should optimise.. it needs a dynamic array.
You should have a Vec in the self struct you use as cache, so you can clear (cheap) and refill every frame.

Reply:
That said if it's a one off.. and not supposed to run every frame we can keep it but add a comment about why you allow this `collect`

Reply:
> Btw not sure how compiler should optimise.. it needs a dynamic array.

`chunks_exact` returns an iterator that implement both `TrustedLen` and `ExactSizeIterator` traits. With that knowledge, `collect()` can use specialization and pre-allocate the memory knowing the exact size.

Reply:
#420